### PR TITLE
Rk

### PR DESCRIPTION
**Pull request by romakoifman**

### CODE REVIEW

File: src/SpanContainer.cpp [hunk 1]
```diff
@@ -1,47 +1,66 @@
...
 
-    SpanContainer::~SpanContainer(){
-        delete m_span;
-    }
+    SpanContainer::SpanContainer(){}
+    SpanContainer::~SpanContainer(){}
 
-    Span& SpanContainer::GetSpan() const{
-        return *m_span;
+    const Span* SpanContainer::TopSpan() const{
```
Comment (Dudi119) [inline src/SpanContainer.cpp]: TopSpan should only return a reference to the current span, if the list is empty, an exception seems appropriate.

File: src/SpanContainer.cpp [hunk 2]
```diff
@@ -1,47 +1,66 @@
...
+        if (GetSpans()->empty()){
+            return nullptr;
+        }
+        return GetSpans()->front();
     }
 
-    SpanContainer::SpanContainer(){}
+    const Span::SpanHeader* SpanContainer::GetRootHeader() const{
```
Comment (Dudi119) [inline src/SpanContainer.cpp]: No real need for root.

File: src/SpanContainer.cpp [hunk 3]
```diff
@@ -1,47 +1,66 @@
...
-        swap(m_span, span);
-        delete span;
+    void SpanContainer::PushSpan(Span* span) {
+        GetSpans()->push_front(span);
+        if(span->IsRootSpan())
+            m_rootHeader.reset( new Span::SpanHeader(span->GetHeader()));
     }
 
-    unique_ptr<Span> SpanContainer::ReleaseSpan(){
+   Span* SpanContainer::PopSpan(){
```
Comment (Dudi119) [inline src/SpanContainer.cpp]: Should return unique_ptr to reflect transition of ownership. if empty, throw an exception.

File: src/Sampler.cpp [hunk 4]
```diff
@@ -20,11 +20,9 @@ namespace cppkin
 
 #endif
 
-    void Sampler::AdvanceSampler() {
-        currentSample = atomic_fetch_add(&sampler, 1) + 1;
-    }
-    bool Sampler::ShouldSample() {
+    bool Sampler::AdvanceSampler() {
```
Comment (Dudi119) [inline src/Sampler.cpp]: I think the logic should be separated in order to get rid of root and head below.

File: src/cppkin.h [hunk 5]
```diff
@@ -14,87 +14,101 @@
...
+        return;
+    uint_fast64_t id = cppkin::Trace::Instance().GenerateTraceID();
+    std::unique_ptr<cppkin::Span> span = std::move(cppkin::Trace::Instance().CreateSpan(operationName, id));
+    cppkin::SpanContainer::Instance().PushSpan(span.release());
+}
+
+inline void CppKin::CreateSpan(const char* operationName){
+    cppkin::SpanContainer& spanContainer = cppkin::SpanContainer::Instance();
+    const cppkin::Span::SpanHeader* parentHeader = spanContainer.GetRootHeader();
+    if(!parentHeader) // No parent available. Span always has to have a parent. Skipping
```
Comment (Dudi119) [inline src/cppkin.h]: Question the current cycle and not the header, it doesn't seems right we question one time the cycle and the next the existence of the header while the logic of the previous already exists.

File: src/cppkin.h [hunk 6]
```diff
@@ -14,87 +14,101 @@
...
+        return;
+
+    const cppkin::Span* currentSpan = spanContainer.TopSpan();
```
Comment (Dudi119) [inline src/cppkin.h]: Same as previous comment.

File: src/cppkin.h [hunk 7]
```diff
@@ -14,87 +14,101 @@
...
+    if(currentSpan) {
+        parentHeader = &currentSpan->GetHeader();
+    }
+
+    uint_fast64_t id = cppkin::Trace::Instance().GenerateSpanID();
+    std::unique_ptr<cppkin::Span> span = cppkin::Trace::Instance().CreateSpan(operationName, parentHeader->TraceID, parentHeader->ID, id);
```
Comment (Dudi119) [inline src/cppkin.h]: transition the value category to xvalue (move) seems also legit here.

File: src/cppkin.h [hunk 8]
```diff
@@ -14,87 +14,101 @@
...
+    spanContainer.PushSpan(span.release());
+}
+
+inline void CppKin::JoinSpan(const cppkin::Span::SpanHeader& spanHeader) {
+
+    if(cppkin::Sampler::AdvanceSampler()){
```
Comment (Dudi119) [inline src/cppkin.h]: Join should only advance the sampler if the current span is a trace.

File: src/cppkin.h [hunk 9]
```diff
@@ -14,87 +14,101 @@
...
+    //    }
+
+    //    cppkin::EncoderContext##encodingType context;
+    //    cppkin::Encoder<EncodingType::encodingType>::Serialize(context, span.GetHeader());
+    //    return std::move(context.ToString());
+
+    return std::string();
+}
+
+static void SetTracingHeader(char* data, size_t length) {
```
Comment (Dudi119) [inline src/cppkin.h]: use const char* instead.

File: src/SpanContainer.cpp (hunk 1 above)
Comment (romakoifman) [reply src/SpanContainer.cpp]: It depends. Lets discuss tomorrow. If it is the only comment please approve/merge

File: src/cppkin.h [hunk 10]
```diff
@@ -14,87 +14,96 @@
...
+    static inline void JoinSpan(const cppkin::Span::SpanHeader& spanHeader) {
+
+        if (!cppkin::Sampler::ShouldSample())
+            return;
+
+        std::unique_ptr<cppkin::Span> span;
+        if(spanHeader.ParentIDSet)
+             span = std::move(cppkin::Trace::Instance().CreateSpan(spanHeader.Name.c_str(), spanHeader.TraceID, spanHeader.ParentID, spanHeader.ID));
+        else
+            span = std::move(cppkin::Trace::Instance().CreateSpan(spanHeader.Name.c_str(), spanHeader.TraceID));
```
Comment (Dudi119) [inline src/cppkin.h]: When joining a trace, should we set the tls sample count to be valid or not according to what it used to be in prrvious service?

**Pull request merged**